crates_io_tarball/src/lib.rs

```diff
@@ -89,25 +92,46 @@ pub fn process_tarball<R: Read>(
             ));
         }
 
-        if entry_path == vcs_info_path {
-            let mut contents = String::new();
-            entry.read_to_string(&mut contents)?;
-            vcs_info = CargoVcsInfo::from_contents(&contents).ok();
-        } else if entry_path == manifest_path || entry_path == manifest_path_lower {
-            // Try to extract and read the Cargo.toml from the tarball, silently
-            // erroring if it cannot be read.
-            let mut contents = String::new();
-            entry.read_to_string(&mut contents)?;
-            manifest = Some({
+        if entry_path.parent() == Some(pkg_root) {
+            let entry_file = entry_path.file_name().unwrap_or_default();
+            if entry_file == ".cargo_vcs_info.json" {
```

Comment:
this makes it apply to a `foo-1.2.3/a/b/c/.cargo_vcs_info.json` file too, while previously we were only reading the top level file